##### Copyright 2024 Google LLC.

In [ ]:
#@title Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Gemini API Python quickstart

<table class="tfo-notebook-buttons" align="left">
  <td>
    <a target="_blank" href="https://ai.google.dev/tutorials/quickstart_colab"><img src="https://ai.google.dev/static/site-assets/images/docs/notebook-site-button.png" height="32" width="32" />View on Google AI</a>
  </td>
  <td>
    <a target="_blank" href="https://colab.research.google.com/github/google/generative-ai-docs/blob/main/site/en/tutorials/quickstart_colab.ipynb"><img src="https://www.tensorflow.org/images/colab_logo_32px.png" />Run in Google Colab</a>
  </td>
  <td>
    <a target="_blank" href="https://github.com/google/generative-ai-docs/blob/main/site/en/tutorials/quickstart_colab.ipynb"><img src="https://www.tensorflow.org/images/GitHub-Mark-32px.png" />View source on GitHub</a>
  </td>
</table>

This tutorial shows you how to get started with the Gemini API using the Python SDK.

## Prerequisites

You can run this tutorial in Google Colab, which doesn't require additional environment configuration.

Alternatively, to complete this quickstart locally, see the Python guidance in [Get started with the Gemini API](https://ai.google.dev/tutorials/quickstart).

## Install the SDK

The Python SDK for the Gemini API is contained in the [`google-generativeai`](https://pypi.org/project/google-generativeai/) package. Install the dependency using pip:

` pip install -U google-generativeai grpcio `

## Import the package and configure the service with your API key: 
To use the Gemini API, you'll need an API key. If you don't already have one, create a key in Google AI Studio.

<a class="button" href="https://aistudio.google.com/app/apikey" target="_blank" rel="noopener noreferrer">Get an API key</a>

In Colab, add the key to the secrets manager under the "🔑" in the left panel. Give it the name `GOOGLE_API_KEY`. Then pass the key to the SDK:

In [ ]:
import os
import google.generativeai as genai

# id using Colab
# from google.colab import userdata
# GOOGLE_API_KEY=userdata.get('GOOGLE_API_KEY')
# genai.configure(api_key=GOOGLE_API_KEY)

# local
genai.configure(api_key=os.environ['GOOGLE_API_KEY'])

## Initialize the Generative Model

Before you can make any API calls, you need to initialize the Generative Model.

In [5]:
model = genai.GenerativeModel('gemini-1.5-flash')

## Generate text

In [9]:
sample_pdf = genai.upload_file("./OnePageOneTable.pdf")
response = model.generate_content(["Extract the table from this PDF. The first line or two might be the title. The first few lines in the table might have merged cells so the table may have more columns.", sample_pdf])
print(response.text)

|  | Three months ended June 30, | Six months ended June 30, |
|---|---|---|
|  | 2015 | 2014 | 2015 | 2014 |
| Revenue (notes 5 and 6) | $ 2,231 | $ 2,458 | $ 4,476 | $ 5,105 |
| Costs and expenses (income) |  |  |  |  |
| Cost of sales (notes 5 and 7) | 1,689 | 1,631 | 3,397 | 3,350 |
| General and administrative expenses | 70 | 82 | 137 | 185 |
| Exploration, evaluation and project expenses (note 8) | 97 | 105 | 183 | 205 |
| Impairment charges (note 10B) | 35 | 512 | 40 | 524 |
| Loss on currency translation | 33 | 31 | 31 | 110 |
| Closed mine rehabilitation | (19) | 27 | (11) | 49 |
| Loss (gain) on non-hedge derivatives (note 18D) | 8 | (44) | 11 | (65) |
| Other expense (note 10A) | 32 | 17 | 14 | 36 |
| Income before finance items and income taxes | $ 286 | $ 97 | $ 674 | $ 711 |
| Finance items |  |  |  |  |
| Finance income | 2 | 3 | 4 | 6 |
| Finance costs (note 11) | (194) | (200) | (390) | (401) |
| Income (loss) before income taxes | $ 94 | $ (100) | $ 288 | $ 316 |
| In



|  | 2015 | 2014 | 2015 | 2014 |
|---|---|---|---|---|
| Revenue (notes 5 and 6) | $ 2,231 | $ 2,458 | $ 4,476 | $ 5,105 |
| Costs and expenses (income) |  |  |  |  |
| Cost of sales (notes 5 and 7) | 1,689 | 1,631 | 3,397 | 3,350 |
| General and administrative expenses | 70 | 82 | 137 | 185 |
| Exploration, evaluation and project expenses (note 8) | 97 | 105 | 183 | 205 |
| Impairment charges (note 10B) | 35 | 512 | 40 | 524 |
| Loss on currency translation | 33 | 31 | 31 | 110 |
| Closed mine rehabilitation | (19) | 27 | (11) | 49 |
| Loss (gain) on non-hedge derivatives (note 18D) | 8 | (44) | 11 | (65) |
| Other expense (note 10A) | 32 | 17 | 14 | 36 |
| Income before finance items and income taxes | $ 286 | $ 97 | $ 674 | $ 711 |
| Finance items |  |  |  |  |
| Finance income | 2 | 3 | 4 | 6 |
| Finance costs (note 11) | (194) | (200) | (390) | (401) |
| Income (loss) before income taxes | $ 94 | $ (100) | $ 288 | $ 316 |
| Income tax expense (note 12) | (103) | (123) | (208) | (412) |
| Net income (loss) | $ (9) | $ (223) | $ 80 | $ (96) |
| Attributable to: |  |  |  |  |
| Equity holders of Barrick Gold Corporation | $ (9) | $ (269) | $ 48 | $ (181) |
| Non-controlling interests (note 21) | $ - | $ 46 | $ 32 | $ 85 |
| Earnings (loss) per share data attributable to the equity holders of Barrick Gold Corporation (note 9) |  |  |  |  |
| Net income (loss) |  |  |  |  |
| Basic | $ (0.01) | $ (0.23) | $ 0.04 | $ (0.16) |
| Diluted | $ (0.01) | $ (0.23) | $ 0.04 | $ (0.16) |

In [8]:
sample_pdf = genai.upload_file("./OnePageOneTable2023.pdf")
response = model.generate_content(["Extract the table from this PDF. The first line or two might be the title. The first few lines in the table might have merged cells so the table has more rolls", sample_pdf])
print(response.text)

|  | 2023 | 2022 |
|---|---|---|
| Revenue (notes 5 and 6) | $11,397 | $11,013 |
| Costs and expenses (income) |  |  |
| Cost of sales (notes 5 and 7) | 7,932 | 7,497 |
| General and administrative expenses (note 11) | 126 | 159 |
| Exploration, evaluation and project expenses (notes 5 and 8) | 361 | 350 |
| Impairment charges (notes 10 and 21) | 312 | 1,671 |
| Loss on currency translation | 93 | 16 |
| Closed mine rehabilitation (note 27b) | 16 | (136) |
| Income from equity investees (note 16) | (232) | (258) |
| Other (income) expense (note 9) | (195) | (268) |
| Income before finance items and income taxes | 2,984 | 1,982 |
| Finance costs, net (note 14) | (170) | (301) |
| Income before income taxes | 2,814 | 1,681 |
| Income tax expense (note 12) | (861) | (664) |
| Net income | $1,953 | $1,017 |
| Attributable to: |  |  |
| Equity holders of Barrick Gold Corporation | $1,272 | $432 |
| Non-controlling interests (note 32) | $681 | $585 |
| Earnings (loss) per share data attribut

|  | 2023 | 2022 |
|---|---|---|
| Revenue (notes 5 and 6) | $11,397 | $11,013 |
| Costs and expenses (income) |  |  |
| Cost of sales (notes 5 and 7) | 7,932 | 7,497 |
| General and administrative expenses (note 11) | 126 | 159 |
| Exploration, evaluation and project expenses (notes 5 and 8) | 361 | 350 |
| Impairment charges (notes 10 and 21) | 312 | 1,671 |
| Loss on currency translation | 93 | 16 |
| Closed mine rehabilitation (note 27b) | 16 | (136) |
| Income from equity investees (note 16) | (232) | (258) |
| Other (income) expense (note 9) | (195) | (268) |
| Income before finance items and income taxes | 2,984 | 1,982 |
| Finance costs, net (note 14) | (170) | (301) |
| Income before income taxes | 2,814 | 1,681 |
| Income tax expense (note 12) | (861) | (664) |
| Net income | $1,953 | $1,017 |
| Attributable to: |  |  |
| Equity holders of Barrick Gold Corporation | $1,272 | $432 |
| Non-controlling interests (note 32) | $681 | $585 |
| Earnings (loss) per share data attributable to the equity holders of Barrick Gold Corporation (note 13) |  |  |
| Net income |  |  |
| Basic | $0.72 | $0.24 |
| Diluted | $0.72 | $0.24 |

In [17]:
sample_pdf = genai.upload_file("./output1.pdf")
response = model.generate_content(["Extract table from this PDF as MarkDown. The first line or two might be the title. The first few lines in the table might have merged cells so the table has more rolls.", sample_pdf])
print(response.text)

|  | 2018 | 2017 |
|---|---|---|
| **Revenue (notes 5 and 6)** | $1,790 | $1,993 |
| **Costs and expenses (income)** |  |  |
| Cost of sales (notes 5 and 7) | 1,152 | 1,342 |
| General and administrative expenses | 48 | 72 |
| Exploration, evaluation and project expenses | 73 | 75 |
| Impairment (reversals) charges (notes 9B and 13) | 2 | (1,125) |
| Loss on currency translation (note 9C) | 15 | 3 |
| Closed mine rehabilitation | (9) | 8 |
| Income from equity investees (note 12) | (16) | (11) |
| Gain on non-hedge derivatives | (2) | (4) |
| Other expense (note 9A) | 1 | 2 |
| **Income before finance costs and income taxes** | $526 | $1,631 |
| Finance costs, net | (133) | (150) |
| **Income before income taxes** | $393 | $1,481 |
| Income tax expense (note 10) | (201) | (592) |
| **Net income** | $192 | $889 |
| **Attributable to:** |  |  |
| Equity holders of Barrick Gold Corporation | $158 | $679 |
| Non-controlling interests | $34 | $210 |
| **Earnings per share data attributable 

|  | 2018 | 2017 |
|---|---|---|
| **Revenue (notes 5 and 6)** | $1,790 | $1,993 |
| **Costs and expenses (income)** |  |  |
| Cost of sales (notes 5 and 7) | 1,152 | 1,342 |
| General and administrative expenses | 48 | 72 |
| Exploration, evaluation and project expenses | 73 | 75 |
| Impairment (reversals) charges (notes 9B and 13) | 2 | (1,125) |
| Loss on currency translation (note 9C) | 15 | 3 |
| Closed mine rehabilitation | (9) | 8 |
| Income from equity investees (note 12) | (16) | (11) |
| Gain on non-hedge derivatives | (2) | (4) |
| Other expense (note 9A) | 1 | 2 |
| **Income before finance costs and income taxes** | $526 | $1,631 |
| Finance costs, net | (133) | (150) |
| **Income before income taxes** | $393 | $1,481 |
| Income tax expense (note 10) | (201) | (592) |
| **Net income** | $192 | $889 |
| **Attributable to:** |  |  |
| Equity holders of Barrick Gold Corporation | $158 | $679 |
| Non-controlling interests | $34 | $210 |
| **Earnings per share data attributable to the equity holders of Barrick Gold Corporation (note 8)** |  |  |
| **Net income** |  |  |
| Basic | $0.14 | $0.58 |
| Diluted | $0.14 | $0.58 |


## What's next

To learn more about working with the Gemini API, see the [Python tutorial](https://ai.google.dev/tutorials/python_quickstart).

If you're new to generative AI models, you might want to look at the
[concepts guide](https://ai.google.dev/docs/concepts) and the
[Gemini API overview](https://ai.google.dev/docs/gemini_api_overview).